### PR TITLE
Fix incorrect SIMD temp allocation for Vector256 with AVX2 disabled

### LINKED ISSUE #58295: Vector256<Byte.WithElement(1235410822): RunBasicScenario failed to throw ArgumentOutOfRangeException.
```
export COMPlus_TieredCompilation=1
export COMPlus_JitStress=1
```

```
      Beginning scenario: RunBasicScenario
      Vector256<Byte.WithElement(1235410822): RunBasicScenario failed to throw ArgumentOutOfRangeException.
      
      Beginning scenario: RunReflectionScenario
      Beginning scenario: RunBasicScenario
      Beginning scenario: RunBasicScenario
      Beginning scenario: RunReflectionScenario
      Beginning scenario: RunReflectionScenario
      ERROR!!!-System.Exception: One or more scenarios did not complete as expected.
      
      Ending test case at 8/27/2021 1:28:53 AM
      Beginning test case GetAndWithElement.Byte.7 at 8/27/2021 1:28:53 AM
      Random seed: 20010415; set environment variable CORECLR_SEED to this value to repro
      
      Beginning scenario: RunBasicScenario
      Vector256<Byte.WithElement(1235410822): RunBasicScenario failed to throw ArgumentOutOfRangeException.
      
      Beginning scenario: RunReflectionScenario
      Beginning scenario: RunBasicScenario
      Beginning scenario: RunBasicScenario
      Beginning scenario: RunReflectionScenario
      Beginning scenario: RunReflectionScenario
      ERROR!!!-System.Exception: One or more scenarios did not complete as expected.
```

[console.log](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-bcebfc22f41e48df8c/JIT.HardwareIntrinsics/1/console.ef9e075b.log?sv=2019-07-07&se=2021-09-16T05%3A23%3A22Z&sr=c&sp=rl&sig=E%2BSO4pKlkSswPoBg%2BlNXDnT1UsxbcnQbrdrAElt7Ne4%3D)

### PR DESCRIPTION
**Pull request by BruceForstall**
The NI_Vector256_GetElement intrinsic, in some situations, requires
a stack temporary. With AVX2 disabled, this temporary was getting
allocated as a TYP_SIMD16 instead of a TYP_SIMD32, leading to overwriting
the local variable.

Add a type argument to the temp variable allocation, and allocate the
temp as the largest sized type required by any use.

Fixes #58295

### CODE REVIEW

**Comment by BruceForstall** [comment]
Windows x64 diffs:
```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 222
Total bytes of diff: 190
Total bytes of delta: -32 (-14.41% of base)
Total relative delta: -0.29
    diff is an improvement.
    relative diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
         -16 : 207372.dasm (-14.29% of base)
         -16 : 207371.dasm (-14.55% of base)

2 total files with Code Size differences (2 improved, 0 regressed), 0 unchanged.

Top method improvements (bytes):
         -16 (-14.29% of base) : 207372.dasm - GitHub_22815.Program:test128(byte):bool
         -16 (-14.55% of base) : 207371.dasm - GitHub_22815.Program:test128(ubyte):bool

Top method improvements (percentages):
         -16 (-14.55% of base) : 207371.dasm - GitHub_22815.Program:test128(ubyte):bool
         -16 (-14.29% of base) : 207372.dasm - GitHub_22815.Program:test128(byte):bool

2 total methods with Code Size differences (2 improved, 0 regressed), 0 unchanged.

```

</details>

--------------------------------------------------------------------------------

Linux x64:

```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 978
Total bytes of diff: 931
Total bytes of delta: -47 (-4.81% of base)
Total relative delta: -0.33
    diff is an improvement.
    relative diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
         -16 : 210097.dasm (-14.29% of base)
         -16 : 210096.dasm (-14.29% of base)
         -10 : 198748.dasm (-2.72% of base)
          -5 : 195198.dasm (-1.30% of base)

4 total files with Code Size differences (4 improved, 0 regressed), 0 unchanged.

Top method improvements (bytes):
         -16 (-14.29% of base) : 210097.dasm - GitHub_22815.Program:test128(byte):bool
         -16 (-14.29% of base) : 210096.dasm - GitHub_22815.Program:test128(ubyte):bool
         -10 (-2.72% of base) : 198748.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector128`1[Byte],int):this
          -5 (-1.30% of base) : 195198.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector128`1[Byte],int):this

Top method improvements (percentages):
         -16 (-14.29% of base) : 210097.dasm - GitHub_22815.Program:test128(byte):bool
         -16 (-14.29% of base) : 210096.dasm - GitHub_22815.Program:test128(ubyte):bool
         -10 (-2.72% of base) : 198748.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector128`1[Byte],int):this
          -5 (-1.30% of base) : 195198.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector128`1[Byte],int):this

4 total methods with Code Size differences (4 improved, 0 regressed), 0 unchanged.

```

</details>

--------------------------------------------------------------------------------


These diffs are due to allocating a SIMD16 instead of a SIMD32, so the frame shrinks. We don't have any SPMI collections where AVX2 is disabled.

**Comment by BruceForstall** [comment]
@tannergooding @dotnet/jit-contrib PTAL

@tannergooding We could alternately go with your fix proposed in https://github.com/dotnet/runtime/issues/58295#issuecomment-915419187, although this one has the minor benefit of shrinking the temp when the larger temp is not needed. Thoughts?

**Comment by tannergooding** [comment]
> Thoughts?

I think this fix is fine. I thought we were using `getSIMDInitTempVarNum` in more places and this would be a more complex change to make work like you have.

File: src/coreclr/jit/lsraarm64.cpp [hunk 1]
```diff
@@ -1132,7 +1132,7 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
                 {
                     // If the index is not a constant or op1 is in register,
                     // we will use the SIMD temp location to store the vector.
-                    compiler->getSIMDInitTempVarNum();
+                    compiler->getSIMDInitTempVarNum(compiler->getSIMDVectorType());
```
Comment (tannergooding) [inline src/coreclr/jit/lsraarm64.cpp:1135]:
```suggestion
                    var_types requiredSimdTempType = (intrin.id == NI_Vector64_GetElement) ? TYP_SIMD8 : TYP_SIMD16;
                    compiler->getSIMDInitTempVarNum(requiredSimdTempType);
```

**Review by tannergooding — APPROVED** [review]
LGTM.

**Comment by BruceForstall** [comment]
@tannergooding With your suggested change (included here), there are a few SPMI arm64 asm diffs similar to the x64 ones noted above: TYP_SIMD16 temps shrunk to TYP_SIMD8, thus shrinking frame sizes.

## coreclr_tests.pmi.windows.arm64.checked.mch:

```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 2184
Total bytes of diff: 2176
Total bytes of delta: -8 (-0.37% of base)
Total relative delta: -0.02
    diff is an improvement.
    relative diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file regressions (bytes):
           4 : 196147.dasm (0.54% of base)

Top file improvements (bytes):
          -4 : 196012.dasm (-0.56% of base)
          -4 : 196094.dasm (-1.12% of base)
          -4 : 196182.dasm (-1.05% of base)

4 total files with Code Size differences (3 improved, 1 regressed), 0 unchanged.

Top method regressions (bytes):
           4 ( 0.54% of base) : 196147.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector64`1[Byte],int):this

Top method improvements (bytes):
          -4 (-0.56% of base) : 196012.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector64`1[Byte],int):this
          -4 (-1.12% of base) : 196094.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector64`1[Byte],int):this
          -4 (-1.05% of base) : 196182.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector64`1[Byte],int):this

Top method regressions (percentages):
           4 ( 0.54% of base) : 196147.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector64`1[Byte],int):this

Top method improvements (percentages):
          -4 (-1.12% of base) : 196094.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector64`1[Byte],int):this
          -4 (-1.05% of base) : 196182.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector64`1[Byte],int):this
          -4 (-0.56% of base) : 196012.dasm - HVATests`1[Byte][System.Byte]:checkValues(System.String,System.Runtime.Intrinsics.Vector64`1[Byte],int):this

4 total methods with Code Size differences (3 improved, 1 regressed), 0 unchanged.

```

</details>

--------------------------------------------------------------------------------

## libraries.crossgen2.windows.arm64.checked.mch:

```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 1124
Total bytes of diff: 1124
Total bytes of delta: 0 (0.00% of base)
```
<details>

<summary>Detail diffs</summary>

```


0 total files with Code Size differences (0 improved, 0 regressed), 3 unchanged.

0 total methods with Code Size differences (0 improved, 0 regressed), 3 unchanged.

```

</details>

--------------------------------------------------------------------------------

**Comment by tannergooding** [comment]
Should this also add an outerloop leg that will catch this kind of failure in the future?

**Comment by BruceForstall** [comment]
> Should this also add an outerloop leg that will catch this kind of failure in the future?

I'm testing a PR (https://github.com/dotnet/runtime/pull/58822) to add JitStress=1 and JitStress=2 to the "runtime-coreclr jitstress-isas-x86" pipeline. We can discuss it over there.

**Pull request merged**